### ISSUE TITLE
Deferreds should support 'yield from'

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8087]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)| @hawkowl reported|
|-|-|
|Trac ID|trac#8087|
|Type|enhancement|
|Created|2015-11-05 14:10:54Z|
|Branch|https://github.com/twisted/twisted/tree/8087-yield-from|

This is a Python 3.4 feature, that seems like it might make it nicer to write coroutines-style code.

<details><summary>Searchable metadata</summary>

```
trac-id__8087 8087
type__enhancement enhancement
reporter__hawkowl hawkowl
priority__normal normal
milestone__Python_3_x Python-3.x
branch__8087_yield_from 8087-yield-from
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1446732654505809 1446732654505809
changetime__1474198049187875 1474198049187875
version__None None
owner__Amber_Brown__hawkowl_____ Amber Brown <hawkowl@...>

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_1"></a>|@hawkowl commented|
|-|-|

(In [46622]) Branching to yield-from-8087.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)|@hawkowl set owner to @hawkowl|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_3"></a>|@markrwilliams commented|
|-|-|

It's such a straightforward patch!  I'd love to see this get merged.

It needs some tests and I think the topfile's content is for another ticket.  If it's helpful and you're OK with it, I can update the topfile, write some tests, and push it all to this branch.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_4"></a>|@hawkowl commented|
|-|-|

Yes, I do need to write some tests :) But I'll do that, I just haven't got to it yet (using PR for the builds mainly)

There is a benchmark I pushed up earlier, this should make it much quicker...

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_5"></a>|@markrwilliams commented|
|-|-|

OK!  I'm happy to review when this is ready (as I'm sure many others are, too).

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_6"></a>|@hawkowl removed owner|
|-|-|

Happy with this! Please review.

Benchmarks, also:

```
Python 3.5 (trunk)
100.4 deferred_await/sec (502 deferred_await in 5 seconds)
N/A deferred_yieldfrom/sec

Python 3.5 (this branch)
190.8 deferred_await/sec (954 deferred_await in 5 seconds)
191.4 deferred_yieldfrom/sec (957 deferred_yieldfrom in 5 seconds)

Python 3.4 (this branch)
Skipping deferred_await : invalid syntax (deferred_await.py, line 7)
257.0 deferred_yieldfrom/sec (1285 deferred_yieldfrom in 5 seconds)
```

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_7"></a>|@Lukasa commented|
|-|-|

Ok, a quick look from me and it seems fine: one note for the topfile. I'll let mark review as well though.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_8"></a>|@hawkowl set owner to @hawkowl<br>@hawkowl set status to `closed`|
|-|-|

In changeset ddb4f67ccfc5b498d806392d7f5a7af63d9fcd93
```
#!CommitTicketReference repository="" revision="ddb4f67ccfc5b498d806392d7f5a7af63d9fcd93"
Merge 8087-yield-from: Allow Deferreds to be yield from'd

Author: hawkowl
Reviewes: markrwilliams, lukasa
Fixes: #8087
```

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_9"></a>|@hawkowl commented|
|-|-|

[mass edit] Removing review from closed tickets.
